### PR TITLE
Refactor, add DTS support for i2c_sam_twihs driver

### PR DESCRIPTION
**Pull request by mnkp**

### CODE REVIEW

File: drivers/i2c/i2c_sam_twihs.c [hunk 1]
```diff
@@ -10,17 +10,18 @@
  * Only I2C Master Mode with 7 bit addressing is currently supported.
  */
 
+#define SYS_LOG_DOMAIN "dev/i2c_sam_twihs"
```
Comment (galak) [inline drivers/i2c/i2c_sam_twihs.c]: any reason this isn't in one of the cleanup patches?

File: dts/arm/atmel/same70.dtsi [hunk 2]
```diff
@@ -31,6 +32,36 @@
 	};
 
 	soc {
+		i2c0: i2c@40018000 {
+			compatible = "atmel,sam-i2c-twihs";
```
Comment (galak) [inline dts/arm/atmel/same70.dtsi]: all the i2c nodes should have:

                        #address-cells = <1>;
                        #size-cells = <0>;

**Review by galak — CHANGES_REQUESTED** [review]
some minor dt related clean ups

File: dts/arm/atmel/same70.dtsi (hunk 2 above)
Comment (mnkp) [reply dts/arm/atmel/same70.dtsi]: Done.

File: drivers/i2c/i2c_sam_twihs.c (hunk 1 above)
Comment (mnkp) [reply drivers/i2c/i2c_sam_twihs.c]: I had to rearrange the order due to `#include "i2c-priv.h"` which is new. `i2c-priv.h` defines `SYS_LOG_LEVEL`.

Comment (galak) [reply drivers/i2c/i2c_sam_twihs.c]: ah, cool

File: drivers/i2c/twihs_sam.c [hunk 3]
```diff
@@ -135,11 +150,10 @@ static void write_msg_start(Twihs *const twihs, struct twihs_msg *msg,
 {
 	/* Set slave address and number of internal address bytes. */
```
Comment (MaureenHelm) [inline drivers/i2c/twihs_sam.c]: Remove "and number of internal address bytes"

File: drivers/i2c/twihs_sam.c [hunk 4]
```diff
@@ -175,6 +187,12 @@ static int twihs_sam_transfer(struct device *dev, struct i2c_msg *msgs,
 		return 0;
 	}
 
+	/* Clear pending interrupts, such as NACK. */
+	(void)twihs->TWIHS_SR;
```
Comment (MaureenHelm) [inline drivers/i2c/twihs_sam.c]: Why the `(void)`? Does reading this register clear the interrupts?

File: dts/arm/sam_e70_xplained.fixup [hunk 5]
```diff
@@ -6,6 +6,25 @@
 
 #define CONFIG_NUM_IRQ_PRIO_BITS	ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
 
+#define CONFIG_I2C_0_BASE_ADDRESS	ATMEL_SAM_I2C_TWIHS_40018000_BASE_ADDRESS
+#define CONFIG_I2C_0_NAME		ATMEL_SAM_I2C_TWIHS_40018000_LABEL
+#define CONFIG_I2C_0_BITRATE		ATMEL_SAM_I2C_TWIHS_40018000_CLOCK_FREQUENCY
+#define CONFIG_I2C_0_IRQ		ATMEL_SAM_I2C_TWIHS_40018000_IRQ_0
+#define CONFIG_I2C_0_IRQ_PRI		ATMEL_SAM_I2C_TWIHS_40018000_IRQ_0_PRIORITY
+#define CONFIG_I2C_0_PERIPHERAL_ID	ATMEL_SAM_I2C_TWIHS_40018000_PERIPHERAL_ID
```
Comment (MaureenHelm) [inline dts/arm/sam_e70_xplained.fixup]: Please insert a newline between instances to improve readability

File: dts/bindings/i2c/atmel,sam-i2c-twihs.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
+---
```
Comment (MaureenHelm) [inline dts/bindings/i2c/atmel,sam-i2c-twihs.yaml]: Add a license header

**Review by MaureenHelm — DISMISSED** [review]

File: drivers/i2c/twihs_sam.c (hunk 4 above)
Comment (mnkp) [reply drivers/i2c/twihs_sam.c]: It is a common practice to put `(void)` in front of what may look like an unnecessary read operation to indicate that we are intentionally ignoring the returned value. In this case reading status register clears all the interrupts. Should I still remove it?

I'll fix other issues.

Comment (MaureenHelm) [reply drivers/i2c/twihs_sam.c]: No, it's fine.

**Comment by mnkp** [comment]
I've fixed remaining issues and rebased on the latest arm branch.

**Review by galak — DISMISSED** [review]

**Review by fallrisk — APPROVED** [review]

**Review by dbkinder — DISMISSED** [review]
+1 for doc change, but there are merge conflicts

**Comment by mnkp** [comment]
Hi @galak, there are some cmake switch related fixes for Atmel SAM SoCs on the master branch which are still missing on arm branch. Could we rebase the arm branch on the latest master, otherwise it's a bit difficult to work. I'll push an updated version of this PR then.

**Comment by galak** [comment]
@mnkp should be updated now.

**Review by galak — APPROVED** [review]

**Comment by mnkp** [comment]
@galak thanks for working on this PR. I am currently traveling and not always able to respond quickly.

**Pull request merged**